async_hooks: cleanup emitBefore and emitAfter in timers and nextTick

AndreasMadsen opened the pull request:
##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
async_hooks, process, timers

**async_hooks: use common emitBefore and emitAfter** 
* Timers and `nextTick` have special `emitBefore` and `emitAfter` functions for
performance reasons. But if the assert in `async_hooks.emitBefore` is
removed there is no difference.

_edit: the checks are already included in [`pushAsyncIds`]( https://github.com/nodejs/node/blob/f06c05cf5a2c2678a76c15adbe7f3acbc9478617/src/env-inl.h#L130L131)._

**async_hooks: require parameter in emitBefore**
* Using `asyncId` as the default `triggerAsyncId` is wrong. The `triggerAsyncId`
can actually never be the `asyncId`.

/cc @nodejs/async_hooks

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8945/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8950/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8953/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8957/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8960/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8962/

trevnorris commented on lib/async_hooks.js:
```diff
@@ -365,21 +365,14 @@ function emitBeforeN(asyncId) {
-  if (asyncId < 0 || triggerAsyncId < 0) {
-    fatalError('before(): asyncId or triggerAsyncId is less than zero ' +
-               `(asyncId: ${asyncId}, triggerAsyncId: ${triggerAsyncId})`);
-  }
```
This is to throw nicely. If the values make it to `AsyncWrap::PushAsyncIds()` then it will abort (i.e. always core dump).

AndreasMadsen replied:
Yes. But does this matter. `asyncId` and `triggerAsyncId` is for sure not a common error. We can keep the check, but then the performance of `nextTick` and the timers won't be exactly the same.

trevnorris replied:
`emitBeforeS` is a (undocumented) user API and is exported via `require()`. The general rule has been that it shouldn't be possible for the application to abort via any method imported via `require()` (this does not hold true for `process.binding()`). But removing this allows `require('async_hooks').emitBefore(-1, -1)` to cause a `CHECK()` failure.

AndreasMadsen replied:
Okay, that makes sense. I will readd the checks later today.

trevnorris commented on lib/internal/process/next_tick.js:
```diff
@@ -149,24 +147,6 @@ function setupNextTick() {
-      emitBefore(asyncId, triggerAsyncId);
-    else
-      pushAsyncIds(asyncId, triggerAsyncId);
-  }
```
History lesson: This was introduced originally because of a crazy method to keep track of the id stack. Which was a lot faster than it is currently, but added an obscene amount of complexity. I hope to simplify that approach in the future and reintroduce it, but seems I overlooked these calls can be removed for the time being.

trevnorris commented on lib/timers.js:
```diff
@@ -149,22 +146,6 @@ exports._unrefActive = function(item) {
-    emitBefore(asyncId, triggerAsyncId);
-  else
-    pushAsyncIds(asyncId, triggerAsyncId);
-}
```
Reason these were here is for the same reason explained in nextTick.

AndreasMadsen commented:
rebased. CI: https://ci.nodejs.org/job/node-test-pull-request/8995/

AndreasMadsen commented:
I reverted the commits that removed the assert and squashed the commits so it isn't confusing. I also rewrote the commit message.

CI: https://ci.nodejs.org/job/node-test-pull-request/9005/

AndreasMadsen commented:
landed in 4e27aa9603a281caf8241b1bba8e0f6d57554c57 and 84f9261efd02c7e4ddb087cadf528b2aaed8809c
